### PR TITLE
Fix true negative computation

### LINKED ISSUE #100: `true_negatives` in `binary_statistics()` is incorrect
The computation of true negatives based on the confusion matrix looks like it's summing only the diagonal of the submatrix instead of the full submatrix of all classes except the class of interest. Since this is supposed to be a binary metric, it should be summing all classes that are not the class of interest that are also not predicted to be the class of interest.

https://github.com/beacon-biosignals/Lighthouse.jl/blob/e829617c5b30654a0c9765390e541e749b7aae33/src/metrics.jl#L73

### PR DESCRIPTION
**Pull request by alexmchan**
Resolves #100

### CODE REVIEW

**Review by ericphanson — COMMENTED** [review]
The changes look good to me. I suspect the failure also occurs on main, and it is a regression stemming from https://github.com/apache/arrow-julia/pull/371 I believe, cc @omus. I wasn't able to quickly make a small reproduction though.

File: test/metrics.jl
Comment (alexmchan) [reply test/metrics.jl:14]: Yeah, @omus came to the same conclusion and is addressing the issue on the arrow side. We'll wait until that's fixed and CI is passing here.

**Comment by hannahilea** [comment]
Depending on how soon that fix is going to land, we may want to pin our Arrow version before the breaking change so that it doesn't block PRs like this one. I can do that in a separate PR if folks want to land this PR ASAP.

**Comment by hannahilea** [comment]
We may also want to backport this fix to v0.14+, for folks who haven't upgraded Legolas versions to 0.5 yet

**Comment by hannahilea** [comment]
(turning it off and on again to trick testing to happen again!)

**Comment by hannahilea** [comment]
Right, looks like even though the fix was merged in (https://github.com/apache/arrow-julia/pull/371) and released as Arrow 4.3.2 (https://github.com/apache/arrow-julia/releases)---which _is_ being used here in these tests---the tests are failing.

**Comment by ericphanson** [comment]
Hm, https://github.com/apache/arrow-julia/pull/371 didn't fix it maybe? cc @omus 
```julia
  MethodError: no method matching zero(::Type{Union{Missing, Array{Int64}}})
  Closest candidates are:
    zero(::Type{Union{Missing, T}}) where T at missing.jl:105
    zero(::Union{Type{P}, P}) where P<:Dates.Period at /opt/hostedtoolcache/julia/1.8.5/x64/share/julia/stdlib/v1.8/Dates/src/periods.jl:53
    zero(::ColorVectorSpace.RGBRGB) at ~/.julia/packages/ColorVectorSpace/QI5vM/src/ColorVectorSpace.jl:425
    ...
  Stacktrace:
    [1] default(T::Type)
      @ ArrowTypes ~/.julia/packages/ArrowTypes/E6ePy/src/ArrowTypes.jl:303
    [2] ArrowTypes.ToArrow(x::Vector{Union{Missing, Array{Int64}}})
      @ ArrowTypes ~/.julia/packages/ArrowTypes/E6ePy/src/ArrowTypes.jl:353
    [3] arrowvector(x::Vector{Union{Missing, Array{Int64}}}, i::Int64, nl::Int64, fi::Int64, de::Dict{Int64, Any}, ded::Vector{Arrow.DictEncoding}, meta::Nothing; dictencoding::Bool, dictencode::Bool, maxdepth::Int64, kw::Base.Pairs{Symbol, Union{Nothing, Bool}, NTuple{4, Symbol}, NamedTuple{(:compression, :largelists, :denseunions, :dictencodenested), Tuple{Nothing, Bool, Bool, Bool}}})
      @ Arrow ~/.julia/packages/Arrow/MJoOf/src/arraytypes/arraytypes.jl:62
    [4] toarrowvector(x::Vector{Union{Missing, Array{Int64}}}, i::Int64, de::Dict{Int64, Any}, ded::Vector{Arrow.DictEncoding}, meta::Nothing; compression::Nothing, kw::Base.Pairs{Symbol, Integer, NTuple{5, Symbol}, NamedTuple{(:largelists, :denseunions, :dictencode, :dictencodenested, :maxdepth), Tuple{Bool, Bool, Bool, Bool, Int64}}})
      @ Arrow ~/.julia/packages/Arrow/MJoOf/src/arraytypes/arraytypes.jl:37
    [5] (::Arrow.var"#145#146"{Dict{Int64, Any}, Bool, Nothing, Bool, Bool, Bool, Int64, Nothing, Vector{Arrow.DictEncoding}, Vector{Type}, Vector{Any}})(col::Vector{Union{Missing, Array{Int64}}}, i::Int64, nm::Symbol)
      @ Arrow ~/.julia/packages/Arrow/MJoOf/src/write.jl:337
    [6] eachcolumn
      @ ~/.julia/packages/Tables/T7rHm/src/utils.jl:70 [inlined]
    [7] toarrowtable(cols::Tables.CopiedColumns{NamedTuple{(:class_labels, :confusion_matrix, :discrimination_calibration_curve, :discrimination_calibration_score, :multiclass_IRA_kappas, :multiclass_kappa, :optimal_threshold, :optimal_threshold_class, :per_class_IRA_kappas, :per_class_kappas, :stratified_kappas, :per_class_pr_curves, :per_class_reliability_calibration_curves, :per_class_reliability_calibration_scores, :per_class_roc_aucs, :per_class_roc_curves, :per_expert_discrimination_calibration_curves, :per_expert_discrimination_calibration_scores, :spearman_correlation, :thresholds), Tuple{Vector{Union{Missing, Vector{String}}}, Vector{Union{Missing, Array{Int64}}}, Vector{Union{Missing, Tuple{Vector{Float64}, Vector{Float64}}}}, Vector{Union{Missing, Float64}}, Vector{Union{Missing, Float64}}, Vector{Union{Missing, Float64}}, Vector{Union{Missing, Float64}}, Vector{Union{Missing, Int64}}, Vector{Union{Missing, Vector{Float64}}}, Vector{Union{Missing, Vector{Float64}}}, Vector{Union{Missing, Vector{NamedTuple{(:per_class, :multiclass, :n), Tuple{Vector{Float64}, Float64, Int64}}}}}, Vector{Union{Missing, Vector{Tuple{Vector{Float64}, Vector{Float64}}}}}, Vector{Union{Missing, Vector{Tuple{Vector{Float64}, Vector{Float64}}}}}, Vector{Union{Missing, Vector{Float64}}}, Vector{Union{Missing, Vector{Float64}}}, Vector{Union{Missing, Vector{Tuple{Vector{Float64}, Vector{Float64}}}}}, Vector{Union{Missing, Vector{Tuple{Vector{Float64}, Vector{Float64}}}}}, Vector{Union{Missing, Vector{Float64}}}, Vector{Union{Missing, NamedTuple{(:ρ, :n, :ci_lower, :ci_upper), Tuple{Float64, Int64, Float64, Float64}}}}, Vector{Union{Missing, Vector{Float64}}}}}}, dictencodings::Dict{Int64, Any}, largelists::Bool, compress::Nothing, denseunions::Bool, dictencode::Bool, dictencodenested::Bool, maxdepth::Int64, meta::Base.ImmutableDict{String, String}, colmeta::Nothing)
      @ Arrow ~/.julia/packages/Arrow/MJoOf/src/write.jl:334
    [8] macro expansion
      @ ~/.julia/packages/Arrow/MJoOf/src/write.jl:195 [inlined]
    [9] macro expansion
      @ ./task.jl:454 [inlined]
   [10] write(writer::Arrow.Writer{IOBuffer}, source::Vector{EvaluationV1})
      @ Arrow ~/.julia/packages/Arrow/MJoOf/src/write.jl:185
   [11] (::Arrow.var"#137#138"{Vector{EvaluationV1}})(writer::Arrow.Writer{IOBuffer})
      @ Arrow ~/.julia/packages/Arrow/MJoOf/src/write.jl:283
   [12] open(::Arrow.var"#137#138"{Vector{EvaluationV1}}, ::Type, ::Vararg{Any}; kwargs::Base.Pairs{Symbol, Any, Tuple{Symbol, Symbol}, NamedTuple{(:file, :metadata), Tuple{Bool, Tuple{Pair{String, String}}}}})
      @ Base ./io.jl:384
   [13] #write#136
      @ ~/.julia/packages/Arrow/MJoOf/src/write.jl:282 [inlined]
   [14] #write_arrow#59
      @ ~/.julia/packages/Legolas/vrwHc/src/tables.jl:117 [inlined]
   [15] write(io_or_path::IOBuffer, table::Vector{EvaluationV1}, sv::Legolas.SchemaVersion{Symbol("lighthouse.evaluation"), 1}; validate::Bool, metadata::Nothing, kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
      @ Legolas ~/.julia/packages/Legolas/vrwHc/src/tables.jl:222
   [16] write
      @ ~/.julia/packages/Legolas/vrwHc/src/tables.jl:195 [inlined]
   [17] roundtrip_row(row::EvaluationV1)
      @ Main ~/work/Lighthouse.jl/Lighthouse.jl/test/runtests.jl:67

```

**Comment by palday** [comment]
@ericphanson 

MWE:
```julia
tbl = (; x=Union{Missing, Array{Int64}}[missing])
io = IOBuffer()
Arrow.write(io, tbl)
```

The insidious bit here is that the *eltype* of the column is `Union{Missing, Array{Int64}}`.

Forcing the eltype to be a 1D vector helps:

```julia
tbl = (; x=Union{Missing, Vector{Int64}}[missing])
io = IOBuffer()
Arrow.write(io, tbl)
```

So I suspect it's the union with an abstract type that's causing the problem.

**Comment by hannahilea** [comment]
Thanks, @palday! Once #102 lands, THIS PR will be unblocked. phew!

**Review by hannahilea — APPROVED** [review]

**Comment by hannahilea** [comment]
huh---thought a close/open would trigger CI. github ci is currently in a bad state, so i'll try again in the morning.

**Comment by hannahilea** [comment]
Woo! Finally! :D

**Pull request closed without merging**
